### PR TITLE
Makes Wasp smarter when writing over existing generated files

### LINKED ISSUE #209: Make Wasp smarter when writing over existing generated files (e.g. delete redundant(old) files)
Wasp compiles source and generates code in outDir.
But, it does not clear the ourDir first -> instead it just generates the code over whatever is already there, meaning there might be some left-over files that shouldn't exist any more.
This wasn't a problem so far, since redundant files were just not used by the rest of the code.
However, it starting being a problem with .env file, because if you remove .env from Wasp project, the old one remains in the generated project, and env vars from it area loaded while they shouldn't be.

For now we did a quick fix -> we just delete .env file from the generated directory before we generate new code.

Advanced approach would be to, before actually writing file drafts to disk, generate a list of files that are to be created together with their checksums. Then, we can use that list to compare it to the previous list (that we generated the last time we generated code and we persisted it in helper file next to the generated code) and we can do fancy stuff like:
- Delete all the files that were on the previous list but are not any more.
- Not write files that have not changed (that have the same checksum as before), to avoid unnecessary writes to disk.

This means that we need a way to generate this list before actually writing the file drafts to the disk.
I think the best way to do this is to add additional method to Writeable, that is somewhat of a "dry run" method -> it does not write anything to disk, instead it tells us path of where it wants to write + the checksum of the file that it would write. It would probably need to do some IO operations like reading for this, but it will not write anything. This can then be used to construct the list, and then actual `write` method can be used to finally execute the drafts that need to be executed (many of them will not be needed to be executed because they will not change).

Once we do this, we can probably remove `preCleanup` method from the ServerGenerator.hs.

### PR DESCRIPTION
**Pull request by shayneczyzewski**
# Description

Wasp compiles source and generates code in its out dir. However, it does not clear that directory first. This means it overwrites what is there (potentially writing _many_ unchanged files), as well as does not remove files that should be deleted. Previously we had manual `preCleanup` steps that removed redundant files we knew about-`.env`, for example.

This change adds (path, checksum) tracking for all files written as `FileDraft`s to know what is on disk, and uses this to make decisions to either:
- ignore unchanged `FileDraft`s,
- write new/updated `FileDraft`s,
- or remove stale Files or Dirs.

This means we can now remove any `preCleanup` steps. It also means any time we update a `FileDraft` resource outside the normal generation process, we must also invalidate that item in the cache. Therefore, since DB migration does this out of band when the user invokes via the CLI, we have to remove the migrations dir from the checksum file when doing so.

Note: One bit of trickiness we had to deal with was the fact that we could be operating on a file-based `FileDraft` or a directory-based `FileDraft`. However, any generic FS ops typically still have to check for and differentiate between files/dirs.

If you'd like to see the benefits in action, try watching the output directory with `fswatch` or some equivalent.

Closes #209

## Type of change

Please select the option(s) that is more relevant.

- [x] Code cleanup
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

**Comment by shayneczyzewski** [comment]
@Martinsos I added the ability to remove an entry from the Generator maintained checksum file and did so for migrations dir here when migrating: https://github.com/wasp-lang/wasp/pull/486/commits/205407972b4f6f43e94a5a470b6cbd72f099cc39

I went remove instead of update to keep the logic clean and simple, and to make the normal generation process the thing that does all the checksum generation fun.

One other nice property is I made the interface for `readChecksumFile` and `writeChecksumFile` operate on the same types of entities `RelPathsToChecksums`. Lemme know if it seems reasonable, thanks! 👍🏻

File: waspc/src/Wasp/Generator/WriteFileDrafts.hs [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+        error $
+          "Found different file path type! Expected one of: ["
+            ++ fileFsEntityLabel
+            ++ ","
+            ++ dirFsEntityLabel
+            ++ "]. This should never happen!"
+
+writeChecksumFile :: Path' Abs (Dir ProjectRootDir) -> RelPathsToChecksums -> IO ()
+writeChecksumFile dstDir relativePathsToChecksums = do
+  let res2 = first fromSpToTypeAndPath <$> relativePathsToChecksums
```
Comment (shayneczyzewski) [inline waspc/src/Wasp/Generator/WriteFileDrafts.hs:141]: Reminder to clean up this `res2` name

**Comment by shayneczyzewski** [comment]
Gave it another pass @Martinsos and made a few small tweaks, but seems pretty decent to me ;) I think we did ok here haha. Feel free to give it a pass as time allows and if you agree we can remove draft status (from title and PR). Thanks! 👍🏻

**Review by Martinsos — COMMENTED** [review]
Lgtm! Left two small comments. When you are happy with those we can open it up for others!

File: waspc/src/Wasp/Generator/DbGenerator/Operations.hs [hunk 2]
```diff
@@ -52,9 +53,12 @@ migrateDevAndCopyToSource dbMigrationsDirInWaspProjectDirAbs genProjectRootDirAb
     ExitFailure code -> return $ Left $ "Migrate (dev) failed with exit code: " ++ show code
 
 finalizeMigration :: Path' Abs (Dir ProjectRootDir) -> Path' Abs (Dir DbMigrationsDir) -> IO (Either String ())
-finalizeMigration genProjectRootDirAbs dbMigrationsDirInWaspProjectDirAbs =
-  copyMigrationsBackToSource genProjectRootDirAbs dbMigrationsDirInWaspProjectDirAbs
-    <* writeDbSchemaChecksumToFile genProjectRootDirAbs (SP.castFile dbSchemaChecksumOnLastMigrateFileProjectRootDir)
+finalizeMigration genProjectRootDirAbs dbMigrationsDirInWaspProjectDirAbs = do
+  -- NOTE: We are updating a managed CopyDirFileDraft outside the normal generation process, so we must invalidate the checksum entry for it.
+  removeFromChecksumFile genProjectRootDirAbs [Right $ SP.castDir $ dbRootDirInProjectRootDir SP.</> dbMigrationsDirInDbRootDir]
```
[the remark this answers is not in the document]
Comment (shayneczyzewski) [reply waspc/src/Wasp/Generator/DbGenerator/Operations.hs:58]: Good point, will do! 👍🏻

File: waspc/src/Wasp/Generator/WriteFileDrafts.hs [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+
+deleteFilesAndDirs :: Path' Abs (Dir ProjectRootDir) -> [FileOrDirPathRelativeTo ProjectRootDir] -> IO ()
+deleteFilesAndDirs dstDir filesAndDirs = do
+  let absFileFPs = map (\relSP -> SP.fromAbsFile $ dstDir </> relSP) (lefts filesAndDirs)
+  mapM_ removeFile absFileFPs
+  let absDirFPs = map (\relSP -> SP.fromAbsDir $ dstDir </> relSP) (rights filesAndDirs)
+  mapM_ removeDirectoryRecursive absDirFPs
+
+-- | Writes .waspinfo, which contains some basic metadata about how/when wasp generated the code.
+writeDotWaspInfo :: Path' Abs (Dir ProjectRootDir) -> IO ()
```
[the remark this answers is not in the document]
Comment (shayneczyzewski) [reply waspc/src/Wasp/Generator/WriteFileDrafts.hs:171]: Yeah I agree, seems we pulled it over a bit prematurely. I'll put it back into Generator 👍🏻

File: waspc/src/Wasp/Generator/WriteFileDrafts.hs [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+fileDraftsToWriteAndFilesToDelete (Just existingFilePathsToChecksums) fileDraftsWithChecksums =
+  let fileDrafts = fst <$> fileDraftsWithChecksums
+   in ( getNewFileDrafts existingFilePathsToChecksums fileDrafts
+          ++ getChangedFileDrafts existingFilePathsToChecksums fileDraftsWithChecksums,
+        getRedundantGeneratedFiles existingFilePathsToChecksums fileDrafts
+      )
+
+getNewFileDrafts :: RelPathsToChecksums -> [FileDraft] -> [FileDraft]
+getNewFileDrafts existingFilePathsToChecksums fileDrafts =
+  filter (\draft -> isNothing $ lookup (getDstPath draft) existingFilePathsToChecksums) fileDrafts
```
Comment (faassen) [inline waspc/src/Wasp/Generator/WriteFileDrafts.hs:83]: The lookup looks through the entire array, which may not scale well with a lot of file drafts (linear scaling, making this a nested loop, quadratic scaling). Consider using a Map for existingFilePathsToChecksums and then Map.lookup.

**Review by faassen — CHANGES_REQUESTED** [review]
This review focuses on performance and the use of `Data.Map`. `Data.Map` makes performance much better without making the code more difficult, let's use it!

Consider a project with 5000 files (that's large but far from unrealistic; one project I was involved in had  over 5500). In the worse case scenario the quadratic use of `lookup` and `find` to look things up in the checksum would result in 25 million operations performed in `getNewFileDrafts`, `getChangedFileDrafts`, `getRedundantGeneratedFiles`. Now the worst case won't happen but it's still quadratic scaling, with millions of operations.

The fix is simple: convert `existingFilePathsToChecksums` using `Data.map` once, and using `Map lookup` and `Map member` instead of array `lookup` and `find`. This code will now scale linearly with the amount of files.

File: waspc/src/Wasp/Generator/WriteFileDrafts.hs (hunk 4 above)
Comment (shayneczyzewski) [reply waspc/src/Wasp/Generator/WriteFileDrafts.hs:83]: Agreed this could become perf issue as project sizes increase.

**Comment by shayneczyzewski** [comment]
> This review focuses on performance and the use of `Data.Map`. `Data.Map` makes performance much better without making the code more difficult, let's use it!
> 
> Consider a project with 5000 files (that's large but far from unrealistic; one project I was involved in had over 5500). In the worse case scenario the quadratic use of `lookup` and `find` to look things up in the checksum would result in 25 million operations performed in `getNewFileDrafts`, `getChangedFileDrafts`, `getRedundantGeneratedFiles`. Now the worst case won't happen but it's still quadratic scaling, with millions of operations.
> 
> The fix is simple: convert `existingFilePathsToChecksums` using `Data.map` once, and using `Map lookup` and `Map member` instead of array `lookup` and `find`. This code will now scale linearly with the amount of files.

Good eye and suggestions, @faassen! For other readers, Martijn and I chatted on Discord and then paired on this for a bit.

I explained originally `StrongPath`s were not compatible with non-`List` types so Martin and I punted. It is now `Hashable`, but that puts us in IO if we want a proper hash table (right?). Martijn and I did tests on this branch (https://github.com/wasp-lang/wasp/tree/shayne-martijn-use-maps-and-sets with this change https://github.com/wasp-lang/wasp/commit/708716b63bc5d01931c594c26893db297214ecf0 ... has a breaking test and needs cleanup, but gives idea) by making `StrongPath`s `Ord` instances. This lets us use `Data.Map` (which is not constant, but `log n` still better than `n`) and should speed it up.

I was a bit hesitant about the immediate need ("use List until it becomes a problem"), as I assumed for hundreds (or even low thousands) of files it would be negligible, but if we can put this problem to bed for good we might as well sooner rather than later. And after seeing the code I think semantically it does express things a bit clearer, with a performance gain at scale. Note: It also indicated it would be good if we had some profiling in our project.

So, while our current project sizes should not notice, if we get to that 50k project file size (which would be HUGE for Wasp adoption) this will matter. I believe the thinking is it's probably fine to ignore this for now until `StrongPath` gets `Ord`, then we can swap out `List` for `Map/Set`, right @faassen?

P.S. I did a quick test here: https://gist.github.com/shayneczyzewski/a1eb17f402ad22198e792d092032a586 showing where the difference starts breaking down some (about 5k-10k you may start to notice).

File: waspc/e2e-test/test-outputs/waspBuild-golden/files.manifest [hunk 5]
```diff
@@ -1,4 +1,5 @@
 waspBuild/.wasp/build/.dockerignore
+waspBuild/.wasp/build/.waspchecksums
```
Comment (matijaSos) [inline waspc/e2e-test/test-outputs/waspBuild-golden/files.manifest:2]: aha, so we have checksums in build/ folder too, when a user runs `wasp build`? I guess this way we avoid extra work during the build phase as well?

File: waspc/e2e-test/test-outputs/waspBuild-golden/waspBuild/.wasp/build/.waspchecksums [hunk 6]
```diff
@@ -0,0 +1,303 @@
+[
+    [
+        [
+            "file",
```
Comment (matijaSos) [inline waspc/e2e-test/test-outputs/waspBuild-golden/waspBuild/.wasp/build/.waspchecksums:4]: if there was a folder, it would be "dir"?

File: waspc/src/Wasp/Generator.hs [hunk 7]
```diff
@@ -42,7 +41,6 @@ writeWebAppCode spec dstDir sendMessage = do
   case generatorResult of
     Left generatorErrors -> return (generatorWarnings, toList generatorErrors)
     Right fileDrafts -> do
-      preCleanup spec dstDir
```
Comment (matijaSos) [inline waspc/src/Wasp/Generator.hs:45]: nice, I always enjoy a good code deletion :D

File: waspc/src/Wasp/Generator/DbGenerator/Operations.hs [hunk 8]
```diff
@@ -52,9 +53,14 @@ migrateDevAndCopyToSource dbMigrationsDirInWaspProjectDirAbs genProjectRootDirAb
     ExitFailure code -> return $ Left $ "Migrate (dev) failed with exit code: " ++ show code
 
 finalizeMigration :: Path' Abs (Dir ProjectRootDir) -> Path' Abs (Dir DbMigrationsDir) -> IO (Either String ())
-finalizeMigration genProjectRootDirAbs dbMigrationsDirInWaspProjectDirAbs =
-  copyMigrationsBackToSource genProjectRootDirAbs dbMigrationsDirInWaspProjectDirAbs
-    <* writeDbSchemaChecksumToFile genProjectRootDirAbs (SP.castFile dbSchemaChecksumOnLastMigrateFileProjectRootDir)
+finalizeMigration genProjectRootDirAbs dbMigrationsDirInWaspProjectDirAbs = do
```
Comment (matijaSos) [inline waspc/src/Wasp/Generator/DbGenerator/Operations.hs:56]: huh this one was a bit hard to follow for me :D Here's what I understood:
- first it removes migrations dir from checksum file, so we'll have to write to the disk again during the next compilation -> why is that important? I saw the note, but couldn't manage to construct the "bad scenario" that we want to avoid in my head
- then it copies the migrations folder from out/generated dir to the wasp project dir
- it writes db schema checksum to some separate file? So we have another checksum file just for db schema, or is that the part of the original .waspchecksums file?

File: waspc/src/Wasp/Generator/FileDraft.hs [hunk 9]
```diff
@@ -40,6 +40,16 @@ instance Writeable FileDraft where
   write dstDir (FileDraftCopyDirFd draft) = write dstDir draft
   write dstDir (FileDraftTextFd draft) = write dstDir draft
 
+  getChecksum (FileDraftTemplateFd draft) = getChecksum draft
```
Comment (matijaSos) [inline waspc/src/Wasp/Generator/FileDraft.hs:43]: oh some sweet TH can't wait to get sprinkled all over here :D

cat_TH

File: waspc/src/Wasp/Generator/FileDraft/CopyFileDraft.hs [hunk 10]
```diff
@@ -47,3 +43,10 @@ instance Writeable CopyFileDraft where
     where
       srcFilePath = SP.fromAbsFile $ _srcPath draft
       absDraftDstPath = absDstDirPath </> _dstPath draft
+
+  -- NOTE: We are reading file here, but then also again later in the "write", so
+  --   we are reading file at least twice! Any way to avoid it?
+  --   Idea: We could cache it in an mvar in the CopyFileDraft.
```
Comment (matijaSos) [inline waspc/src/Wasp/Generator/FileDraft/CopyFileDraft.hs:49]: nice catch!

File: waspc/src/Wasp/Generator/WriteFileDrafts.hs [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+import System.IO.Error (isDoesNotExistError)
+import UnliftIO.Exception (catch, throwIO)
+import Wasp.Generator.Common (ProjectRootDir)
+import Wasp.Generator.FileDraft (FileDraft, write)
+import Wasp.Generator.FileDraft.Writeable (FileOrDirPathRelativeTo, Writeable (getChecksum, getDstPath))
+import Wasp.Util (Checksum)
+
+-- | Writes given file drafts to disk, in the provided destination directory.
+-- Also makes sure to remove any redundant file drafts that have been left on the disk from before.
+-- It is smart when writing, so it doesn't write file drafts that are already written on the disk from before.
```
Comment (matijaSos) [inline waspc/src/Wasp/Generator/WriteFileDrafts.hs:30]: maybe we should reflect this smartness in the name of the function as well? Otherwise it sounds a bit unexpected that `writeFileDrafts` is not writing all the file drafts. Maybe it could be `writeFileDraftsSmart`, or `writeFileDraftsUseChecksums`, or `ensureCorrectFilesAreWrittenToDisk`, just throwing the ideas here :D

File: waspc/src/Wasp/Generator/WriteFileDrafts.hs [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+  case maybePathsToChecksums of
+    -- If checksums file is missing/corrupted, we delete all of the generated code
+    -- since we are not sure what is the state on the disk and don't want redundant files
+    -- to remain after we write the new file drafts.
+    -- Normally we delete redundant left-over files based on the checksum file, but since we don't have it,
+    -- the only way to be sure is just to delete everything.
+    Nothing -> do
+      removeDirectoryRecursive (SP.fromAbsDir dstDir)
+        `catch` \e -> unless (isDoesNotExistError e) $ throwIO e
+    -- We remove checksum file even if it is valid, so in case we fail in the middle writing on the disk
```
Comment (matijaSos) [inline waspc/src/Wasp/Generator/WriteFileDrafts.hs:43]: nice! :D Now whoever reads our source code will successfully pass the interview :D

**Review by matijaSos — COMMENTED** [review]
amazing to see this come to life! :D Made a first pass, I am still bit on a high-level so asked some questions to clarify things for myself :)

File: waspc/e2e-test/test-outputs/waspBuild-golden/waspBuild/.wasp/build/.waspchecksums (hunk 6 above)
Comment (Martinsos) [reply waspc/e2e-test/test-outputs/waspBuild-golden/waspBuild/.wasp/build/.waspchecksums:4]: Yes!

File: waspc/src/Wasp/Generator/FileDraft.hs (hunk 9 above)
Comment (Martinsos) [reply waspc/src/Wasp/Generator/FileDraft.hs:43]: Oh no :D!

File: waspc/src/Wasp/Generator/WriteFileDrafts.hs (hunk 11 above)
Comment (Martinsos) [reply waspc/src/Wasp/Generator/WriteFileDrafts.hs:30]: Yeah @shayneczyzewski and I discussed it but had hard time figuring out a name!
It writes files that need writing, and removes files that are redundant.
So how to capture that in a name? Tricky!
Hmmm maybe `writeGeneratedProjectToDisk`?
Or `ensureProjectIsGeneratedOnTheDisk`?
Not sure hm!

File: waspc/e2e-test/test-outputs/waspBuild-golden/files.manifest (hunk 5 above)
Comment (shayneczyzewski) [reply waspc/e2e-test/test-outputs/waspBuild-golden/files.manifest:2]: Ah, e2e manifests still helping already :D yeah we just used the `outDir` and did not differentiate/care which out dir it was.

File: waspc/src/Wasp/Generator/WriteFileDrafts.hs (hunk 11 above)
Comment (shayneczyzewski) [reply waspc/src/Wasp/Generator/WriteFileDrafts.hs:30]: I'm not the best at name golf admittedly, and this is hard since it both _conditionally_ writes _and_ deletes... 🤔 what we are wanting to convey is the in-memory representation gets materialized and matches the on-disk representation- that they are in sync. So maybe `ensureGeneratedProjectIsConsistent`, or `synchronizeFileDraftsWithDisk`? Ah I dunno on this one sorry :/

Comment (Martinsos) [reply waspc/src/Wasp/Generator/WriteFileDrafts.hs:30]: Ooooo `synchronizeFileDraftsWithDisk` is nice :D!

Comment (shayneczyzewski) [reply waspc/src/Wasp/Generator/WriteFileDrafts.hs:30]: haha thanks! @matijaSos does `synchronizeFileDraftsWithDisk` sound better to you?

**Review by matijaSos — APPROVED** [review]
LGTM from my side!

**Review by faassen — APPROVED** [review]
My comments were all about the algorithms, which we have now separated from this PR. So I'm approved.
